text/0035-emergency-restarts.md

```diff
@@ -0,0 +1,116 @@
+
+The criteria for validity of a block at height `h` should also take this into account, i.e. a block at height `h` is only valid if its global state hash is the result of applying the above steps to the global state specified at height `h-1`.
+
+Note that this has some desired properties:
+- We can specify any change to the state at any height without breaking the continuity of the chain. The first block after the restart is just like any other regular block, which means that eg. chain synchronization after the restart will work just like before the restart.
+- If the last safe block before the restart was at height `h`, but there were also blocks `h+1...h+n` that were finalized, they will become invalid after an upgrade point is specified at height `h+1`. Since the upgrade point wasn't included in the chainspec before the restart, the global state of the old block `h+1` won't satisfy the validity condition after an upgrade is added and the network is restarted. This means that all the blocks finalized after `h` will be discarded after restarting with the new chainspec and the network will resume operation at height `h+1`.
+
+### Consensus
+
+On the consensus side, it is important for an era change to happen at the boundary where the emergency restart happened. One of the reasons is that the fix for the issue that caused the crash could involve changes to the consensus logic, which could render the state of the last era invalid. This means that the conditions defining a switch block have to change as well: a switch block will be a block that either contains the `EraEnd` structure, or that has height `h-1` if there is an upgrade point specified for height `h`.
+
+### The structure of an upgrade point
+
+Currently, an upgrade point supports defining some new config values. We need it to also be able to specify some transforms that should be applied to the global state. This could be realized in multiple ways:
+
+1. Specifying the modifications as direct key-value pairs to be written to the global state (potentially overwriting the values under existing keys).
+2. Specifying arbitrary WASM to be executed, which would result in the transforms to be applied.
+3. For some special use cases, specifying changes in other ways, eg. support for changes to the validator set to be specified using an `accounts.csv`-like file.
+
+In this CEP, we suggest a combination of options 2 and 3.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The main thing required is to extend the `components::chainspec_loader::chainspec::UpgradePoint` structure to make it possible to specify arbitrary modifications to the global state.
+
+The best way of doing that is left as an unresolved question. We suggest adding the following two fields:
+
+```rust
+struct UpgradePoint {
+    // ... other fields
+    /// the paths to arbitrary WASM to be executed, in the specified order
+    upgrade_wasm_paths: Vec<String>,
```

Comment:
we don't need this concept any more, post EE-1050.

Reply:
But that's not about the WASM for system contracts - it's for some code that would perform the changes to the global state that we require. These could also be deploys we'd want to replay from blocks that would become orphaned after a restart.

Reply:
That doesn't make sense to me. If you want to change global state directly the map of key, value concept is much more direct. If you want to execute deploys, you would call existing exec & commit from the contract_runtime component rather than try to burden the upgrade_commit method with the machinery to do that for you.

Reply:
> If you want to change global state directly the map of key, value concept is much more direct & wasmless.

No disagreement here, my concern is how easy it would be to:
1. Prepare such an upgrade, so that it does what we want it to do (eg. reset the set of validators to some agreed upon one),
2. Audit such an upgrade. The validators would probably want to read the set of changes that will be applied and understand what effect that will have on the system. This can be hard to achieve with just a list of the keys and values that will be changed.

> If you want to execute deploys, you would extend the contract_runtime component to compose the EE's existing exec & commit methods using the actual deploys from the deploy store (nearly or perhaps exactly the same as we all ready do for normal execution) rather than try to burden the upgrade_commit method with the machinery to do that for you and without taking on extra complexity in supporting some special cased wasm.

If I understand you correctly, you mean that if we want to use WASM, we shouldn't be passing the deploys to `commit_upgrade`, but rather execute them and apply their results in some other place in the code, like the contract_runtime component? If yes, that sounds reasonable to me.